### PR TITLE
Update convert

### PR DESCRIPTION
**Pull request by exomishra**

### CODE REVIEW

File: deepsearch/cps/client/components/documents.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    """
+    An instance of DocumentConversionResult is generated when document conversion is requested.
+    """
+
+    def __init__(
+        self,
+        proj_key: str,
+        task_ids: list,
+        statuses: List[str],
+        source_file: Optional[Path] = None,
```
Comment (PeterStaar-IBM) [inline deepsearch/cps/client/components/documents.py:25]: `file` -> `path`

File: deepsearch/documents/cli/main.py [hunk 2]
```diff
@@ -49,12 +42,13 @@ def convert(
     """
     typer.echo(WELCOME)
```
Comment (PeterStaar-IBM) [inline deepsearch/documents/cli/main.py:43]: keep log clean

File: deepsearch/documents/core/convert.py [hunk 3]
```diff
@@ -15,6 +15,7 @@
 from deepsearch.cps.client.api import CpsApi
 
 from .common_routines import ERROR_MSG, progressbar_padding
+from .utils import download_url
```
Comment (PeterStaar-IBM) [inline deepsearch/documents/core/convert.py:19]: let's import a small class that does the url black magic.

File: deepsearch/documents/core/main.py [hunk 4]
```diff
@@ -25,10 +28,18 @@ def convert_documents(
     url : string [OPTIONAL]
     For converting a document from the web, please provide its url.
 
-    local_file : path [OPTIONAL]
+    source_file : path [OPTIONAL]
```
Comment (PeterStaar-IBM) [inline deepsearch/documents/core/main.py:31]: file -> path

**Review by PeterStaar-IBM — COMMENTED** [review]
look at the comments

File: deepsearch/documents/core/input_process.py [hunk 5]
```diff
@@ -1,173 +1,118 @@
...
+                root_dir=Path(tmpdir),
+                progress_bar=progress_bar,
+            )
+            statuses = check_status_running_tasks(
+                api=api,
+                cps_proj_key=cps_proj_key,
+                task_ids=task_ids,
+                progress_bar=progress_bar,
+            )
+        if cli_use == False:
```
Comment (cau-git) [inline deepsearch/documents/core/input_process.py:57]: Sorry, we can't do this. This is in a library function and it must not need _any_ knowledge that a CLI even exists.
The CLI implementation should simply use the library classes/method

File: deepsearch/cps/client/components/documents.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        source_path: Optional[Path] = None,
+        source_urls: Optional[List[str]] = None,
+    ) -> None:
+        self.proj_key = proj_key
+        self.task_ids = task_ids
+        self.statuses = statuses
+        self._source_path = source_path
+        self._source_urls = source_urls
+
+    def download_json(self, result_dir: Path, progress_bar=False):
```
Comment (cau-git) [inline deepsearch/cps/client/components/documents.py:36]: Suggests this downloads _one_ JSON. But that is not what it looks like in the code. Please revise the use of plural and singular.

File: deepsearch/cps/client/components/documents.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        Download result of an individual conversion task.
+
+        Input
+        -----
+        result_dir : path
+            local directory where converted documents are stored
+        progress_bar: boolean, optional (default = False)
+            shows progress bar is True
+        """
+        urls = get_download_url(cps_proj_key=self.proj_key, task_ids=[self.task_id])
```
Comment (cau-git) [inline deepsearch/cps/client/components/documents.py:119]: Again, watch the use of plural and singular. This suggests that multiple `urls` can be returned, but the expectation is _one_ document.

File: deepsearch/documents/core/utils.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+from typing import Any, List
+from .common_routines import progressbar_padding
+from deepsearch.cps.client.api import CpsApi
+
+
+class URLNavigator:
+    def __init__(self, api: CpsApi) -> None:
+        self.api = api
+        self.url_host = self.api.client.swagger_client.configuration.host
+        self.url_linked_ccs = (
```
Comment (cau-git) [inline deepsearch/documents/core/utils.py:18]: This looks like dangerous hackery, why are you reverse engineering the linked_ccs URL? It is well defined and one can easily use an URL parser to properly get the base host, path, etc.

I see that in other places you already used `urllib.parse.urlparse`

File: deepsearch/documents/core/main.py [hunk 9]
```diff
@@ -25,32 +30,50 @@ def convert_documents(
...
     For converting local files, please provide absolute path to file or to directory
     containing multiple files.
 
-    NOTE: Either url or local_file should be supplied.
+    progress_bar : Boolean (default is False in code, True in CLI)
+    Show progress bar for processing, submitting, converting input and
+    downloading converted document.
+
+    cli_use : Boolean (default is False in code, True in CLI)
```
Comment (cau-git) [inline deepsearch/documents/core/main.py:41]: As explained, a `cli_use` argument is a no-go.

File: deepsearch/documents/core/main.py [hunk 10]
```diff
@@ -25,32 +30,50 @@ def convert_documents(
...
     # check required inputs are present
-    if url is None and local_file is None:
+    if url is None and source_path is None:
         raise ValueError(
             "No input provided. Please provide either a url or a local file for conversion."
         )
-    elif url is not None and local_file is None:
-        if isinstance(url, str):
+    elif url is not None and source_path is None:
+        if urllib.parse.urlparse(url).scheme in ("http", "https"):
```
Comment (cau-git) [inline deepsearch/documents/core/main.py:58]: What is this supposed to do? Accept URLs with `file://` or such?
Any valid URL must be using `http` or `https` protocol handler, so please remove the extra if/else.

**Review by cau-git — CHANGES_REQUESTED** [review]
@exomishra Thanks! This is not bad so far, but there are a couple of things that we need to change.

1. Any use of the parameter `cli_use`, which is passed around in _library_ functions and implements alternative behaviour, must be eliminated. Library functions must have no knowledge that a CLI exists. This is an absolutely critical aspect of proper code design. 
2. I could not clearly see it because the diff is complex, but please check carefully that _not a single `print()`_ is ever used in a library function. Using `print()` is allowed for CLI only methods.
3. There is an inconsistent use of singular and plural, for example:
```
urls = get_download_url(cps_proj_key=self.proj_key, task_ids=self.task_ids)
```
4. Further comments below.

**Review by cau-git — APPROVED** [review]
🚢 Time to ship!

Below remarks to be kept on record for next iterations.

File: deepsearch/documents/core/convert.py [hunk 11]
```diff
@@ -209,21 +149,28 @@ def send_files_for_conversion(
...
 
-def check_status_running_tasks(api: CpsApi, cps_proj_key: str, task_ids) -> List[str]:
+def check_status_running_tasks(
+    cps_proj_key: str, task_ids, api: Optional[CpsApi] = None, progress_bar=False
+) -> List[str]:
     """
-    Check status of multiple running tasks and display progress with progress bar.
+    Check status of multiple running tasks and optionally display progress with progress bar.
     """
+    if api is None:
```
[the remark this answers is not in the document]
Comment (exomishra) [reply deepsearch/documents/core/convert.py:158]: Should we create an issue to remind/track this source of problem?

**Pull request merged**